File: Python/Práctica/Substring.py
import random

def factorial(n):
	return 1 if (n == 1 or n == 0) else factorial(n-1)*n

def permutation(words):
	permu_words = []
	same_size = True
	size_item = len(words[0])

	for i in words:
		if len(i) == size_item:
			size_item = len(i)
		else:
			same_size = False
			break
	
	if same_size:
		l = 0
		word = ""
		r = factorial(len(words))
		while(l<r):
			for i in range(len(words)):
				w = random.choice(words)
				if w not in word:
					word+=w
				else:
					w = random.choice(words)

			if word not in permu_words and len(word)==len(words)*size_item:
				permu_words.append(word)
				l+=1
			word = ""
	return permu_words

def find_substring(string, words):
	permu_words = permutation(words)
	positions = []
	for word in permu_words:
		if string.find(word)!=-1:
			positions.append(string.find(word))
	return positions

print(find_substring("barfoothefoobarman", ["bar", "foo"]))

File: C++/archivos cpp/Substring.cpp
#include<iostream>
#include<algorithm>
#include<random>
#include<vector>
using namespace std;

int factorial(int n){
  return (n == 1 || n == 0) ? 1 : factorial(n - 1) * n;
}

vector<string> permutation(vector<string> words){
	vector<string> permu_words;
	bool same_size = true;
	int size = words[0].length(), temp_size; // 
	
	// CHECK IF ALL WORDS HAVE THE SAME SIZE
	for(int i=1; i<words.size(); i++){
		if(words[i].length() == size){
			temp_size = size;
			size = words[i].length();
		}else{
			same_size = false;
			break;
		}
	}

	// IF ALL WORDS HAVE THE SAME SIZE SO IT'LL GENERATE ALL PERMUTATION
	if(same_size){
		int l = 0;
		int r = factorial(words.size()); // IT'S THE QUANTITY OF PERMUTATIONS NECESSARY
		random_device rd;
		mt19937 g(rd());
		string word, temp;
		while(l<r){
			shuffle(words.begin(), words.end(), g); // MIX THE WORDS'S VECTOR
			for(auto& x : words){
				word+=x; // GENERATING COMBINATION
			}
			if (word!=temp && temp!=""){
				// IN CASE TO REPEATS MIX NEWLY THE WORDS'S VECTOR
				if(find(permu_words.begin(), permu_words.end(), word) != permu_words.end()){
					shuffle(words.begin(), words.end(), g);
				}
				// IF NOT, ADD THE NEW PERMUTATION IN THE VECTOR
				else{
					permu_words.push_back(word);
					l++;
				}
			}
			// UPDATE THE VARIABLES
			temp = word;
			word = "";
		}
	}
	return permu_words;
}


vector<int> findSubstring(string s, vector<string>& words){
	vector<string> permu_words = permutation(words);
	vector<int> positions;
	// FINDING THE POSITIONS OF EACH WORD IF THIS IS FOUND INTO MAIN WORD
	for(string i : permu_words){
		if(s.find(i) != std::string::npos){
			positions.push_back(s.find(i));
		} 
	}

	return positions;
}

int main(int argc, char const *argv[]){
	vector<string> words = {"bar","foo"};
	vector<int> positions_words = findSubstring("barfoothefoobarman", words);
	// SHOWING POSITIONS
	for(auto& i : positions_words){
		cout<<i<<" ";
	}
	return 0;
}

